developers/opencv.md

```diff
@@ -0,0 +1,251 @@
+            [
+                'videotestsrc ! decodebin', \
+                '! videoconvert ! video/x-raw,format=(string)BGR ! videoconvert',
+                '! appsink'
+            ]
+
+        Args:
+            config (list, optional): Gstreamer pileline description list
+        """
+
+        if not config:
+            config = \
+                [
+                    'videotestsrc ! decodebin',
+                    '! videoconvert ! video/x-raw,format=(string)BGR ! videoconvert',
+                    '! appsink'
+                ]
+
+        command = ' '.join(config)
+        self.video_pipe = Gst.parse_launch(command)
+        self.video_pipe.set_state(Gst.State.PLAYING)
+        self.video_sink = self.video_pipe.get_by_name('appsink0')
+
+    @staticmethod
+    def gst_to_opencv(sample):
+        """Transform byte array into np array
+
+        Args:
+            sample (TYPE): Description
+
+        Returns:
+            TYPE: Description
+        """
+        buf = sample.get_buffer()
+        caps = sample.get_caps()
+        array = np.ndarray(
+            (
+                caps.get_structure(0).get_value('height'),
+                caps.get_structure(0).get_value('width'),
+                3
+            ),
+            buffer=buf.extract_dup(0, buf.get_size()), dtype=np.uint8)
+        return array
+
+    def frame(self):
+        """ Get Frame
+
+        Returns:
+            iterable: bool and image frame, cap.read() output
+        """
+        return self._frame
+
+    def frame_available(self):
+        """Check if frame is available
+
+        Returns:
+            bool: true if frame is available
+        """
+        return type(self._frame) != type(None)
+
+    def run(self):
+        """ Get frame to update _frame
+        """
+
+        self.start_gst(
+            [
+                self.video_source,
+                self.video_codec,
+                self.video_decode,
+                self.video_sink_conf
+            ])
+
+        self.video_sink.connect('new-sample', self.callback)
+
+    def callback(self, sink):
+        sample = sink.emit('pull-sample')
+        new_frame = self.gst_to_opencv(sample)
+        self._frame = new_frame
+
+        return Gst.FlowReturn.OK
+
+
+if __name__ == '__main__':
+    # Create the video object
+    # Add port= if is necessary to use a different one
+    video = Video()
+
+    while True:
+        # Wait for the next frame
+        if not video.frame_available():
+            continue
+
+        frame = video.frame()
+        cv2.imshow('frame', frame)
+        if cv2.waitKey(1) & 0xFF == ord('q'):
+            break
+```
+
+### Companion
+
+Before running any program in the companion board, keep in mind that the hardware of a SBC may not run with high performance or with real-time requirements.
+
+It's possible to get some frames while the stream is in progress, however this can result in some delays in the top side computer.
+
+#### Get a frame and save it
+
```

Comment:
Do we need to stop the streaming process to release the camera first?

Reply:
Nops.